### PR TITLE
Enable QoS overrides

### PR DESCRIPTION
**Pull request by audrow**
This is adding QoS overrides, which were added in https://github.com/ros2/rclcpp/pull/1408.

This PR relies on
* [x] https://github.com/ros2/message_filters/pull/56

### CODE REVIEW

**Review by SteveMacenski — APPROVED** [review]

**Comment by audrow** [comment]
@SteveMacenski, the required PR has been merged in. I believe this is ready to merge.

File: include/robot_localization/ros_robot_localization_listener.hpp [hunk 1]
```diff
@@ -49,6 +49,27 @@
 namespace robot_localization
 {
 
+namespace detail
+{
+rclcpp::SubscriptionOptions
+get_subscription_options_for_qos_override()
```
Comment (SteveMacenski) [inline include/robot_localization/ros_robot_localization_listener.hpp]: Is a total list of the override options something **not** provided by the rclcpp API? Don't you think this function should be in `rclcpp`?

File: src/ros_filter.cpp [hunk 2]
```diff
@@ -2002,14 +2002,28 @@ void RosFilter<T>::initialize()
     tf2::toMsg(tf2::Transform::getIdentity());
 
   // Position publisher
+  rclcpp::PublisherOptions publisher_options;
+  publisher_options.qos_overriding_options = rclcpp::QosOverridingOptions {{
```
Comment (SteveMacenski) [inline src/ros_filter.cpp]: There's too much redundancy of this list. Can this not be included in a utils file that all of the uses of the list can be grabbed from? I think the "right" answer is to have the `get_default_overridable_options()` (or something similar) function in rclcpp itself, this seems like something **a ton** of people will need / want in hundreds of files / subscribers / publishers.

**Review by SteveMacenski — CHANGES_REQUESTED** [review]

File: src/ros_filter.cpp (hunk 2 above)
Comment (audrow) [reply src/ros_filter.cpp]: Good call. I'll make a PR to `rclcpp` with something similar.

Comment (jacobperron) [reply src/ros_filter.cpp]: There is the method [`with_default_policies`](https://github.com/ros2/rclcpp/blob/0251f70fe8a952fae242e6dcdc5b11bc4c11dc76/rclcpp/include/rclcpp/qos_overriding_options.hpp#L141) that I suggest we call here instead (unless someone feels strongly about exposing configuration points for QoS settings other than history, depth, and reliability).

**Comment by audrow** [comment]
I'm now using `rclcpp::QosOverridingOptions::with_default_policies()`, which removes much duplication. It doesn't enable all of the QoS overrides, but after some discussion with the ROS 2 team, it was determined that not all the options are desirable: https://github.com/ros-perception/image_pipeline/pull/651#discussion_r635608091.

**Review by SteveMacenski — APPROVED** [review]

**Comment by SteveMacenski** [comment]
@audrow Some change here or in your message filters PR is failing to build robot_localization with galactic https://github.com/ros2/message_filters/pull/56.

https://build.ros2.org/job/Gbin_uF64__robot_localization__ubuntu_focal_amd64__binary/7/console

```

23:08:22 /tmp/binarydeb/ros-galactic-robot-localization-3.2.2/src/ros_robot_localization_listener.cpp: In constructor ‘robot_localization::RosRobotLocalizationListener::RosRobotLocalizationListener(rclcpp::Node::SharedPtr, rclcpp::SubscriptionOptions)’:
23:08:31 /tmp/binarydeb/ros-galactic-robot-localization-3.2.2/src/ros_robot_localization_listener.cpp:90:26: error: no matching function for call to ‘message_filters::Subscriber<nav_msgs::msg::Odometry_<std::allocator<void> > >::Subscriber(rclcpp::Node::SharedPtr&, const char [14], rmw_qos_profile_t&, rclcpp::SubscriptionOptions&)’
23:08:31    90 |   tf_listener_(tf_buffer_)
23:08:31       |                          ^
23:08:31 In file included from /tmp/binarydeb/ros-galactic-robot-localization-3.2.2/include/robot_localization/ros_robot_localization_listener.hpp:37,
23:08:31                  from /tmp/binarydeb/ros-galactic-robot-localization-3.2.2/src/ros_robot_localization_listener.cpp:51:
23:08:31 /opt/ros/galactic/include/message_filters/subscriber.h:129:3: note: candidate: ‘message_filters::Subscriber<M>::Subscriber() [with M = nav_msgs::msg::Odometry_<std::allocator<void> >]’
23:08:31   129 |   Subscriber() = default;
23:08:31       |   ^~~~~~~~~~
23:08:31 /opt/ros/galactic/include/message_filters/subscriber.h:129:3: note:   candidate expects 0 arguments, 4 provided
23:08:31 /opt/ros/galactic/include/message_filters/subscriber.h:121:3: note: candidate: ‘message_filters::Subscriber<M>::Subscriber(rclcpp::Node*, const string&, rmw_qos_profile_t) [with M = nav_msgs::msg::Odometry_<std::allocator<void> >; std::string = std::__cxx11::basic_string<char>; rmw_qos_profile_t = rmw_qos_profile_t]’
23:08:31   121 |   Subscriber(rclcpp::Node* node, const std::string& topic, const rmw_qos_profile_t qos = rmw_qos_profile_default)
23:08:31       |   ^~~~~~~~~~
23:08:31 /opt/ros/galactic/include/message_filters/subscriber.h:121:3: note:   candidate expects 3 arguments, 4 provided
23:08:31 /opt/ros/galactic/include/message_filters/subscriber.h:116:3: note: candidate: ‘message_filters::Subscriber<M>::Subscriber(rclcpp::Node::SharedPtr, const string&, rmw_qos_profile_t) [with M = nav_msgs::msg::Odometry_<std::allocator<void> >; rclcpp::Node::SharedPtr = std::shared_ptr<rclcpp::Node>; std::string = std::__cxx11::basic_string<char>; rmw_qos_profile_t = rmw_qos_profile_t]’
23:08:31   116 |   Subscriber(rclcpp::Node::SharedPtr node, const std::string& topic, const rmw_qos_profile_t qos = rmw_qos_profile_default)
23:08:31       |   ^~~~~~~~~~
23:08:31 /opt/ros/galactic/include/message_filters/subscriber.h:116:3: note:   candidate expects 3 arguments, 4 provided
23:08:31 /tmp/binarydeb/ros-galactic-robot-localization-3.2.2/src/ros_robot_localization_listener.cpp:90:26: error: no matching function for call to ‘message_filters::Subscriber<geometry_msgs::msg::AccelWithCovarianceStamped_<std::allocator<void> > >::Subscriber(rclcpp::Node::SharedPtr&, const char [22], rmw_qos_profile_t&, rclcpp::SubscriptionOptions&)’
23:08:31    90 |   tf_listener_(tf_buffer_)
23:08:31       |                          ^
23:08:31 In file included from /tmp/binarydeb/ros-galactic-robot-localization-3.2.2/include/robot_localization/ros_robot_localization_listener.hpp:37,
23:08:31                  from /tmp/binarydeb/ros-galactic-robot-localization-3.2.2/src/ros_robot_localization_listener.cpp:51:
23:08:31 /opt/ros/galactic/include/message_filters/subscriber.h:129:3: note: candidate: ‘message_filters::Subscriber<M>::Subscriber() [with M = geometry_msgs::msg::AccelWithCovarianceStamped_<std::allocator<void> >]’
23:08:31   129 |   Subscriber() = default;
23:08:31       |   ^~~~~~~~~~
23:08:31 /opt/ros/galactic/include/message_filters/subscriber.h:129:3: note:   candidate expects 0 arguments, 4 provided
23:08:31 /opt/ros/galactic/include/message_filters/subscriber.h:121:3: note: candidate: ‘message_filters::Subscriber<M>::Subscriber(rclcpp::Node*, const string&, rmw_qos_profile_t) [with M = geometry_msgs::msg::AccelWithCovarianceStamped_<std::allocator<void> >; std::string = std::__cxx11::basic_string<char>; rmw_qos_profile_t = rmw_qos_profile_t]’
23:08:31   121 |   Subscriber(rclcpp::Node* node, const std::string& topic, const rmw_qos_profile_t qos = rmw_qos_profile_default)
23:08:31       |   ^~~~~~~~~~
23:08:31 /opt/ros/galactic/include/message_filters/subscriber.h:121:3: note:   candidate expects 3 arguments, 4 provided
23:08:31 /opt/ros/galactic/include/message_filters/subscriber.h:116:3: note: candidate: ‘message_filters::Subscriber<M>::Subscriber(rclcpp::Node::SharedPtr, const string&, rmw_qos_profile_t) [with M = geometry_msgs::msg::AccelWithCovarianceStamped_<std::allocator<void> >; rclcpp::Node::SharedPtr = std::shared_ptr<rclcpp::Node>; std::string = std::__cxx11::basic_string<char>; rmw_qos_profile_t = rmw_qos_profile_t]’
23:08:31   116 |   Subscriber(rclcpp::Node::SharedPtr node, const std::string& topic, const rmw_qos_profile_t qos = rmw_qos_profile_default)
23:08:31       |   ^~~~~~~~~~
```

Is this something you can patch?  It looks like you may have removed (?) the default constructor

**Comment by audrow** [comment]
Hey @SteveMacenski, I believe that this is because the message filters PR didn't get into Galactic (it was done after the freeze). I'll ask the ROS boss what to do.

**Comment by audrow** [comment]
Since the change wasn't meant to make it in Galactic, I made a PR reverting the change on the `galactic` branch: #670.

**Pull request merged**